### ISSUE TITLE
[FEATURE REQ] [KeyVault.Keys.KeyClient] How to use ManagedIdentity Access Token with the KeyClient?

### ISSUE THREAD

**Issue by hkuadithya** [Azure/azure-sdk-for-net#17194]
## Query/Question
We're from Azure Container Registry team. This is a question regarding `Azure.Security.KeyVault.Keys`.

We're trying to implement KeyVault VNET/firewall bypass using User managed identity. Here's the [documentation](https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingUserAssignedForCmkPolicy.html) shared by the Identity team. For this, we need to mint a special access token and pass the token to the `KeyVault` client to bypass KeyVault VNET/firewall (we have first party appId and certificate).

It's fairly straightforward to pass the Managed identity token using `AuthenticationCallback` in the [KeyVaultClient](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.keyvault.keyvaultclient.-ctor?view=azure-dotnet-legacy#Microsoft_Azure_KeyVault_KeyVaultClient__ctor_Microsoft_Azure_KeyVault_KeyVaultClient_AuthenticationCallback_System_Net_Http_DelegatingHandler___) constructor. NuGet Package: `Microsoft.Azure.KeyVault`
```
// Use an Async Delegate here for authenticationCallback and pass the access token to KeyVaultClient.

var client = new KeyVaultClient(authenticationCallback: async (authority, resource, _) => 
{
      // Use client assertion with MSAL to obtain access token
      var identityAuthResult = await app.AcquireTokenForClient(...) 
      return identityAuthResult.AccessToken;
});

// Next get the Key from KeyVault
var result = await client.GetKeyAsync("https://adupadhykv.vault-int.azure-int.net/keys/validKey/");
```
---
However, `Microsoft.Azure.KeyVault` NuGet package is now [Deprecated](https://www.nuget.org/packages/Microsoft.Azure.KeyVault).
>This package has been deprecated. The latest libraries to interact with the Azure KeyVault service are: Azure.Security.KeyVault.Keys

I'm not sure how to pass the access token to the new and recommended [KeyClient](https://docs.microsoft.com/en-us/dotnet/api/azure.security.keyvault.keys.keyclient?view=azure-dotnet) of `Azure.Security.KeyVault.Keys` package.

* The constructor takes in 2 arguments, i.e. `VaultUri` and `TokenCredential`. 

* I tried looking into various [TokenCredentials](https://docs.microsoft.com/en-us/dotnet/api/azure.core.tokencredential?view=azure-dotnet) such as `AuthorizationCodeCredential`, `AzureCliCredential`, `ManagedIdentityCredential` etc. 

**How can I use the access token with the new `KeyClient`? Any code references/documentation to help me get started would be greatly appreciated. I tried looking around a bit but couldn't figure out.**

--- 

**[StackOverflow thread](https://stackoverflow.microsoft.com/questions/232587) for tracking this.**

## Environment:
 - Name and version of the Library package used:  **Azure.Security.KeyVault.Keys 4.0.0**
 - Hosting platform or OS and .NET runtime version: **Windows, .NET Framework 4.7.2**
 - IDE and version : **VS 2017/19**

**Comment by heaths** [comment]
You can derive TokenCredential to do the same thing in the GetToken and GetTokenAsync overrides. @schaabs may have more details.

**Comment by hkuadithya** [comment]
@heaths do you suggest creating a class which extends `TokenCredential` and implements these methods?
I was wondering if there was a way to use the one of the [existing class](https://docs.microsoft.com/en-us/dotnet/api/azure.core.tokencredential?view=azure-dotnet) to achieve the stated purpose. 

If not, then creating the custom class is the only feasible option.

**Comment by heaths** [comment]
IT doesn't appear that any of the existing classes will do this for you, but it should be straight forward to make a `TokenCredential` that just takes the bearer token you already have and returns it back in `GetToken` and `GetTokenAsync`. You'll have to manage lifetime either way since you retrieved and took ownership of the bearer token.

**Comment by hkuadithya** [comment]
@heaths I've been trying to do that. I just created a class which extends `TokenCredentials` and overrides `GetToken` and `GetTokenAsync`. These methods are just returning the ManagedIdentity AccessToken minted by MSAL.

However, this seems to succeed with `KeyVaultClient` (deprecated) and fails with `KeyClient` (newer).

For more context, here's the code. Let me know if anything is hard to follow:-

##  Testing the Access token (works with old SDK. Fails with new SDK - Bad request 400)
```
               var identityAccessTokenCredential = new IdentityAccessTokenCredential(
                                msalAuthorityUri: msalAuthorityUri,
                                identityClientId: identityClientId,
                                identityTenantId: identityTenantId,
                                identityCertificate: identityCertificate,
                                keyVaultResourceId: keyVaultResourceId,
                                firstPartyAccessToken: firstPartyAuthResult.AccessToken);

                #endregion

                #region MANAGED IDENTITY TOKEN
                
               // OLD SDK 
                var kvClient1 = new KeyVaultClient(authenticationCallback: async (authority, resource, scope) =>
                {
                    var cred = await identityAccessTokenCredential.GetTokenAsync(requestContext: new TokenRequestContext());
                    return cred.Token;
                });
                //  Get Key Works with old SDK here
                var key1 =  await kvClient1.GetKeyAsync("https://adupadhykv.vault-int.azure-int.net/keys/validKey/");

                // NEW SDK
                var kvClient2 = new KeyClient(vaultUri: new Uri(keyVaultUri), credential: identityAccessTokenCredential);
                
                // Get Key fails with new SDK here
                var key2 = await kvClient2.GetKeyAsync("https://adupadhykv.vault-int.azure-int.net/keys/validKey/");

```

## class IdentityAccessTokenCredential : TokenCredential

```
 class IdentityAccessTokenCredential : TokenCredential
        {
            private readonly string msalAuthorityUri;
            private readonly string ...

            protected IdentityAccessTokenCredential()
            {

            }

            public IdentityAccessTokenCredential(
                string msalAuthorityUri,
                string ...)
            {
                this.msalAuthorityUri = ...
            }

            public override AccessToken GetToken(
                TokenRequestContext requestContext,
                CancellationToken cancellationToken = default)
            {

                return AcquireAccessTokenAsync().GetAwaiter().GetResult();
            }

            public override ValueTask<AccessToken> GetTokenAsync(
                TokenRequestContext requestContext,
                CancellationToken cancellationToken = default)
            {
                return AcquireAccessTokenAsync();
            }
           
            private async ValueTask<AccessToken> AcquireAccessTokenAsync()
            {
                  // USE MSAL to mint Managed Identity token.
            }
```

## Error from KeyClient. Fails while calling GetKeyAsync

```
{"Service request failed.\r\nStatus: 400 (Bad Request)\r\n\r\nContent:\r\n<!DOCTYPE html>\r\n<html>\r\n    <head>\r\n        <title>Runtime Error</title>\r\n        <meta name=\"viewport\" content=\"width=device-width\" />\r\n        <style>\r\n         body {font-family:\"Verdana\";font-weight:normal;font-size: .7em;color:black;} \r\n         p {font-family:\"Verdana\";font-weight:normal;color:black;margin-top: -5px}\r\n         b {font-family:\"Verdana\";font-weight:bold;color:black;margin-top: -5px}\r\n         H1 { font-family:\"Verdana\";font-weight:normal;font-size:18pt;color:red }\r\n         H2 { font-family:\"Verdana\";font-weight:normal;font-size:14pt;color:maroon }\r\n         pre {font-family:\"Consolas\",\"Lucida Console\",Monospace;font-size:11pt;margin:0;padding:0.5em;line-height:14pt}\r\n         .marker {font-weight: bold; color: black;text-decoration: none;}\r\n         .version {color: gray;}\r\n         .error {margin-bottom: 10px;}\r\n         .expandable { text-decoration:underline; font-
weight:bold; color:navy; cursor:hand; }\r\n         @media screen and (max-width: 639px) {\r\n          pre { width: 440px; overflow: auto; white-space: pre-wrap; word-wrap: break-word; }\r\n         }\r\n         @media screen and (max-width: 479px) {\r\n          pre { width: 280px; }\r\n         }\r\n        </style>\r\n    </head>\r\n\r\n    <body bgcolor=\"white\">\r\n\r\n            <span><H1>Server Error in '/' Application.<hr width=100% size=1 color=silver></H1>\r\n\r\n            <h2> <i>Runtime Error</i> </h2></span>\r\n\r\n            <font face=\"Arial, Helvetica, Geneva, SunSans-Regular, sans-serif \">\r\n\r\n            <b> Description: </b>An application error occurred on the server. The current custom error settings for this application prevent the details of the application error from being viewed remotely (for security reasons). It could, however, be viewed by browsers running on the local server machine.\r\n            <br><br>\r\n\r\n            <b>Details:</b> To enable the details of this
 specific error message to be viewable on remote machines, please create a &lt;customErrors&gt; tag within a &quot;web.config&quot; configuration file located in the root directory of the current web application. This &lt;customErrors&gt; tag should then have its &quot;mode&quot; attribute set to &quot;Off&quot;.<br><br>\r\n\r\n            <table width=100% bgcolor=\"#ffffcc\">\r\n               <tr>\r\n                  <td>\r\n                      <code><pre>\r\n\r\n&lt;!-- Web.Config Configuration File --&gt;\r\n\r\n&lt;configuration&gt;\r\n    &lt;system.web&gt;\r\n        &lt;customErrors mode=&quot;Off&quot;/&gt;\r\n    &lt;/system.web&gt;\r\n&lt;/configuration&gt;</pre></code>\r\n\r\n                  </td>\r\n               </tr>\r\n            </table>\r\n\r\n            <br>\r\n\r\n            <b>Notes:</b> The current error page you are seeing can be replaced by a custom error page by modifying the &quot;defaultRedirect&quot; attribute of the application&#39;s &lt;customErrors&gt; configuration tag
 to point to a custom error page URL.<br><br>\r\n\r\n            <table width=100% bgcolor=\"#ffffcc\">\r\n               <tr>\r\n                  <td>\r\n                      <code><pre>\r\n\r\n&lt;!-- Web.Config Configuration File --&gt;\r\n\r\n&lt;configuration&gt;\r\n    &lt;system.web&gt;\r\n        &lt;customErrors mode=&quot;RemoteOnly&quot; defaultRedirect=&quot;mycustompage.htm&quot;/&gt;\r\n    &lt;/system.web&gt;\r\n&lt;/configuration&gt;</pre></code>\r\n\r\n                  </td>\r\n               </tr>\r\n            </table>\r\n\r\n            <br>\r\n\r\n    </body>\r\n</html>\r\n\r\n\r\nHeaders:\r\nStrict-Transport-Security: REDACTED\r\nX-Content-Type-Options: REDACTED\r\nCache-Control: private\r\nDate: Tue, 01 Dec 2020 00:46:37 GMT\r\nX-Powered-By: REDACTED\r\nContent-Length: 3420\r\nContent-Type: text/html; charset=utf-8\r\n"}
    Data: {System.Collections.ListDictionaryInternal}
    ErrorCode: null
    HResult: -2146233088
    HelpLink: null
    InnerException: null
    Message: "Service request failed.\r\nStatus: 400 (Bad Request)\r\n\r\nContent:\r\n<!DOCTYPE html>\r\n<html>\r\n    <head>\r\n        <title>Runtime Error</title>\r\n        <meta name=\"viewport\" content=\"width=device-width\" />\r\n        <style>\r\n         body {font-family:\"Verdana\";font-weight:normal;font-size: .7em;color:black;} \r\n         p {font-family:\"Verdana\";font-weight:normal;color:black;margin-top: -5px}\r\n         b {font-family:\"Verdana\";font-weight:bold;color:black;margin-top: -5px}\r\n         H1 { font-family:\"Verdana\";font-weight:normal;font-size:18pt;color:red }\r\n         H2 { font-family:\"Verdana\";font-weight:normal;font-size:14pt;color:maroon }\r\n         pre {font-family:\"Consolas\",\"Lucida Console\",Monospace;font-size:11pt;margin:0;padding:0.5em;line-height:14pt}\r\n         .marker {font-weight: bold; color: black;text-decoration: none;}\r\n         .version {color: gray;}\r\n         .error {margin-bottom: 10px;}\r\n         .expandable { text-decoration:unde
rline; font-weight:bold; color:navy; cursor:hand; }\r\n         @media screen and (max-width: 639px) {\r\n          pre { width: 440px; overflow: auto; white-space: pre-wrap; word-wrap: break-word; }\r\n         }\r\n         @media screen and (max-width: 479px) {\r\n          pre { width: 280px; }\r\n         }\r\n        </style>\r\n    </head>\r\n\r\n    <body bgcolor=\"white\">\r\n\r\n            <span><H1>Server Error in '/' Application.<hr width=100% size=1 color=silver></H1>\r\n\r\n            <h2> <i>Runtime Error</i> </h2></span>\r\n\r\n            <font face=\"Arial, Helvetica, Geneva, SunSans-Regular, sans-serif \">\r\n\r\n            <b> Description: </b>An application error occurred on the server. The current custom error settings for this application prevent the details of the application error from being viewed remotely (for security reasons). It could, however, be viewed by browsers running on the local server machine.\r\n            <br><br>\r\n\r\n            <b>Details:</b> To enable the det
ails of this specific error message to be viewable on remote machines, please create a &lt;customErrors&gt; tag within a &quot;web.config&quot; configuration file located in the root directory of the current web application. This &lt;customErrors&gt; tag should then have its &quot;mode&quot; attribute set to &quot;Off&quot;.<br><br>\r\n\r\n            <table width=100% bgcolor=\"#ffffcc\">\r\n               <tr>\r\n                  <td>\r\n                      <code><pre>\r\n\r\n&lt;!-- Web.Config Configuration File --&gt;\r\n\r\n&lt;configuration&gt;\r\n    &lt;system.web&gt;\r\n        &lt;customErrors mode=&quot;Off&quot;/&gt;\r\n    &lt;/system.web&gt;\r\n&lt;/configuration&gt;</pre></code>\r\n\r\n                  </td>\r\n               </tr>\r\n            </table>\r\n\r\n            <br>\r\n\r\n            <b>Notes:</b> The current error page you are seeing can be replaced by a custom error page by modifying the &quot;defaultRedirect&quot; attribute of the application&#39;s &lt;customErrors&gt; confi
guration tag to point to a custom error page URL.<br><br>\r\n\r\n            <table width=100% bgcolor=\"#ffffcc\">\r\n               <tr>\r\n                  <td>\r\n                      <code><pre>\r\n\r\n&lt;!-- Web.Config Configuration File --&gt;\r\n\r\n&lt;configuration&gt;\r\n    &lt;system.web&gt;\r\n        &lt;customErrors mode=&quot;RemoteOnly&quot; defaultRedirect=&quot;mycustompage.htm&quot;/&gt;\r\n    &lt;/system.web&gt;\r\n&lt;/configuration&gt;</pre></code>\r\n\r\n                  </td>\r\n               </tr>\r\n            </table>\r\n\r\n            <br>\r\n\r\n    </body>\r\n</html>\r\n\r\n\r\nHeaders:\r\nStrict-Transport-Security: REDACTED\r\nX-Content-Type-Options: REDACTED\r\nCache-Control: private\r\nDate: Tue, 01 Dec 2020 00:46:37 GMT\r\nX-Powered-By: REDACTED\r\nContent-Length: 3420\r\nContent-Type: text/html; charset=utf-8\r\n"
    Source: "Azure.Security.KeyVault.Keys"
    StackTrace: "   at Azure.Security.KeyVault.KeyVaultPipeline.<SendRequestAsync>d__27.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Azure.Security.KeyVault.KeyVaultPipeline.<SendRequestAsync>d__21`1.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Azure.Security.KeyVault.Keys.KeyClient.<GetKeyAsync>d__18.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at KeyVaultApp.VnetBypass.<TestKeyVaultWithVNET>d__13.MoveNext() in C:\\Users\\adupadhy\\source\\repos\\KeyVaultApp\\KeyVaultAp
p\\VnetBypass.cs:line 126"
    Status: 400
    TargetSite: {Void MoveNext()}
```

**Comment by heaths** [comment]
The fact you're getting back an HTML error page from ASP.NET tells me something is wrong in your request URL. At least for Key Vault, I know this happens with colons (":") in the URL path. @schaabs do you have some idea here?

Can you include the URL you're using but just fake the hostname/tenant ID (whatever is actually PII) so we can see the full example URL?

**Comment by schaabs** [comment]
@hkuadithya If your running the exact code your specified above I think the problem might be how you're calling `GetKeyAsync`.  The `GetKeyAsync` method on `KeyClient` expects the name of the key as the first parameter.

https://github.com/Azure/azure-sdk-for-net/blob/05d83431e2d6e3876b9da4e3af8a6407ad6b34e8/sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeyClient.cs#L347

However, in your example you're passing in the entire url (also called the key id).

```c#
var key2 = await kvClient2.GetKeyAsync("https://adupadhykv.vault-int.azure-int.net/keys/validKey/");
```

The new `KeyClient` is contructed with the base vault url, and there is no overload for `GetKeyAsync` which takes the full url as there was on the `KeyVaultClient` in previous versions.

**Comment by hkuadithya** [comment]
@schaabs Yes, I was about to comment the same. I started looking at the new SDK source code after @heaths  mentioned that there's something wrong with the request URL.

The reason why I made this mistake is because the Old SDK expects a `KeyIdentifier` string and the new SDK expects the `KeyName` and `KeyVersion` (optional).

## Incorrect code
```
var kvClient2 = new KeyClient(vaultUri: new Uri("https://adupadhykv.vault-int.azure-int.net"), credential: identityAccessTokenCredential);
var key2 = await kvClient2.GetKeyAsync("https://adupadhykv.vault-int.azure-int.net/keys/validKey/");
```

## Correct code
```
var kvClient2 = new KeyClient(vaultUri: new Uri("https://adupadhykv.vault-int.azure-int.net"), credential: identityAccessTokenCredential);
var key2 = await kvClient2.GetKeyAsync("validKey");
```

Thanks. Please close this issue. Also, it might be useful in future to document how to use the AccessToken directly with the KeyVault SDK.
